### PR TITLE
PAR - Thomas BEguin

### PR DESCRIPTION
**Pull request by liloow**

### CODE REVIEW

File: routes/authors.js [hunk 1]
```diff
@@ -1,8 +1,55 @@
 const express = require("express");
 const generateId = require("uuid/v4");
+const authors = require("../data/authors");
+const body = require('body-parser')
```
Comment (yacinehmito) [inline routes/authors.js:4]: I don't think you need `body-parser` here.

File: routes/authors.js [hunk 2]
```diff
@@ -1,8 +1,55 @@
...
+
 
 const router = express.Router();
 
 // Define your endpoints here
 
-module.exports = router;
+router.get('/', (req, res, next) => {
+	const sorter = req.query.sort
+	if (sorter) res.json(authors.map(el => el[sorter]).sort())
```
Comment (yacinehmito) [inline routes/authors.js:15]: If you do that you are only going to send the sorted names, not the authors sorted by names. Still good code however.

File: routes/authors.js [hunk 3]
```diff
@@ -1,8 +1,55 @@
...
+	res.json(authors)
+})
+
+router.get('/=', (req, res, next) => {
```
Comment (yacinehmito) [inline routes/authors.js:19]: I don't understand what you're trying to do.

File: routes/authors.js [hunk 4]
```diff
@@ -1,8 +1,55 @@
...
+	res.json(authors.map(el => el.name))
+})
+
+
+router.get('/:id', (req, res, next) => {
+	res.json(authors.filter(el => el.id === req.params.id))
```
Comment (yacinehmito) [inline routes/authors.js:25]: Good idea. However, filter will return an array, whereas you only want to send an object. You would use the `find` method to send only the first match instead.

By the way, you also need to handle the case where an author doesn't exist.

File: routes/authors.js [hunk 5]
```diff
@@ -1,8 +1,55 @@
...
+})
+
+router.post('/', (req, res, next) => {
+	const { name, website } = req.body
+	const author = {
+		id: generateId,
```
Comment (yacinehmito) [inline routes/authors.js:31]: `generateId` is a function; you need to call it.

File: routes/authors.js [hunk 6]
```diff
@@ -1,8 +1,55 @@
...
+})
+
+router.put('/:id', (req, res, next) => {
+	const { name, website } = req.body
+	const author = {
+		id: req.params.id,
+		name,
+		website,
+	}
+	authors.push(author)
```
Comment (yacinehmito) [inline routes/authors.js:45]: What happens if an author with the same `id` already exists?

File: routes/authors.js [hunk 7]
```diff
@@ -1,8 +1,55 @@
...
+	res.send(author)
+})
+
+router.get('/:id/books', (req, res, next) => {
+	res.json(books.filter(el => el[req.params.id]).map(el => el.title))
```
Comment (yacinehmito) [inline routes/authors.js:50]: You want to filter using the `author` property of the book objects.

File: routes/books.js [hunk 8]
```diff
@@ -1,8 +1,63 @@
...
+		author,
+		title,
+		description
+	}
+	books.push(book)
+	res.json(book)
+})
+
+router.delete('/:id', (req, res, next) => {
+	books.splice(books.indexOf(books.filter(el => el.id === req.params.id)))
```
Comment (yacinehmito) [inline routes/books.js:33]: That's a pretty cool one, provided that you replace `filter` with `find` (as `filter` will return an array).
You could shorten in with `findIndex`.

File: routes/books.js [hunk 9]
```diff
@@ -1,8 +1,63 @@
...
+	res.status(200)
```
Comment (yacinehmito) [inline routes/books.js:34]: Nice thinking. What you want though is `res.sendStatus(200)`. An even better HTTP code would be 202.

File: routes/books.js [hunk 10]
```diff
@@ -1,8 +1,63 @@
...
+})
+
+router.patch('/:id', (req, res, next) => {
+	if (req.body.id) return res.status(403).json({
```
Comment (yacinehmito) [inline routes/books.js:38]: Awesome! 💯

File: routes/books.js [hunk 11]
```diff
@@ -1,8 +1,63 @@
...
+		author: author || null,
+		description: description || null,
+	}
+	const book = books.filter(el => el.id === req.params.id)[0]
+	for (key of Object.keys(model)) {
+		if (key !== null) {
+			book[key] = model[key]
+			console.log(key)
+		}
+	}
```
Comment (yacinehmito) [inline routes/books.js:55]: It's pretty complicated. `Object.assign` would rescue you.

File: routes/books.js [hunk 12]
```diff
@@ -1,8 +1,63 @@
...
+	res.json({ book })
```
Comment (yacinehmito) [inline routes/books.js:56]: You don't need to enclose your object in another new object. Just send it.

File: routes/books.js [hunk 13]
```diff
@@ -1,8 +1,63 @@
...
+
+})
+
+// router.get(`?author=${authorId}`, (req, res, next) => {
+// 	res.json(authors.filter(el => el.id === req.params.id))
+// })
```
Comment (yacinehmito) [inline routes/books.js:62]: Please clean up your code before submitting.

**Review by yacinehmito — APPROVED** [review]
I very much liked the effort made in using a functional style. I dig it. 😉 

Still, I have the feeling that you wrote your handlers without testing them. You are supposed to use Postman to make sure that the endpoints do what you think they would do.

When you are going to write your REST API during the project you will need to be more vigilant.

File: routes/authors.js (hunk 1 above)
Comment (liloow) [reply routes/authors.js:4]: Was sending data in the wrong format, it was one of the things I tried, forgot to remove it

File: routes/authors.js (hunk 2 above)
Comment (liloow) [reply routes/authors.js:15]: Indeed, but that was on purpose

File: routes/authors.js (hunk 4 above)
Comment (liloow) [reply routes/authors.js:25]: Yeah, I ran into some trouble with that in the patch route, forgot to apply the [0] to the other ones

File: routes/authors.js (hunk 5 above)
Comment (liloow) [reply routes/authors.js:31]: :duck:

File: routes/authors.js (hunk 6 above)
Comment (liloow) [reply routes/authors.js:45]: :bomb:

File: routes/authors.js (hunk 7 above)
Comment (liloow) [reply routes/authors.js:50]: Oops

File: routes/books.js (hunk 8 above)
Comment (liloow) [reply routes/books.js:33]: Or just [0] my filter :grin:

File: routes/books.js (hunk 9 above)
Comment (liloow) [reply routes/books.js:34]: Thanks for the tip !

File: routes/books.js (hunk 10 above)
Comment (liloow) [reply routes/books.js:38]: :muscle:

**Pull request closed without merging**